pkg/controllermanager/controller/shoot/shoot_reference_control.go

```diff
@@ -0,0 +1,250 @@
+	var (
+		oldShoot = oldObj.(*gardencorev1beta1.Shoot)
+		newShoot = newObj.(*gardencorev1beta1.Shoot)
+	)
+
+	if refChange(oldShoot, newShoot) || newShoot.DeletionTimestamp != nil && !controllerutils.HasFinalizer(newShoot, gardencorev1beta1.GardenerName) {
+		key, err := cache.MetaNamespaceKeyFunc(newObj)
+		if err != nil {
+			gardenlogger.Logger.Errorf("Couldn't get key for object %+v: %v", newObj, err)
+			return
+		}
+		c.shootReferenceQueue.Add(key)
+	}
+}
+
+func refChange(oldShoot, newShoot *gardencorev1beta1.Shoot) bool {
+	return !apiequality.Semantic.Equalities.DeepEqual(oldShoot.Spec.DNS, newShoot.Spec.DNS)
+}
+
+// FinalizerName is the name of the finalizer used for the reference protection.
+const FinalizerName = "gardener.cloud/reference-protection"
+
+// NewShootReferenceReconciler creates a new instance of a reconciler which checks object references from shoot objects.
+func NewShootReferenceReconciler(l logrus.FieldLogger) reconcile.Reconciler {
+	return &shootReferenceReconciler{
+		logger: l,
+	}
+}
+
+type shootReferenceReconciler struct {
+	ctx          context.Context
+	gardenClient client.Client
+	logger       logrus.FieldLogger
+}
+
+// InjectClient implements `inject.Client`.
+func (r *shootReferenceReconciler) InjectClient(c client.Client) error {
+	r.gardenClient = c
+	return nil
+}
+
+// InjectClient implements `inject.Stoppable`.
+func (r *shootReferenceReconciler) InjectStopChannel(stopCh <-chan struct{}) error {
+	r.ctx = contextutil.FromStopChannel(stopCh)
+	return nil
+}
+
+// Reconcile checks the shoot in the given request for references to further objects in order to protect them the from
+// deletions as long as they are still referenced.
+func (r *shootReferenceReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	shoot := &gardencorev1beta1.Shoot{}
+	if err := r.gardenClient.Get(r.ctx, request.NamespacedName, shoot); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			return reconcile.Result{}, err
+		}
+		r.logger.Infof("Unable to retrieve object %q from store: %v", request.Name, err)
+		return reconcile.Result{}, err
+	}
+
+	r.logger.Infof("[SHOOT REFERENCE CONTROL] %s", request)
+
+	if err := r.reconcileShootReferences(shoot); err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *shootReferenceReconciler) reconcileShootReferences(shoot *gardencorev1beta1.Shoot) error {
+	// Iterate over all user secrets in project namespace and check if they can be released.
+	if err := r.releaseUnreferencedSecrets(shoot); err != nil {
+		return err
+	}
+
+	// Remove finalizer from shoot in case it's being deleted and not handled by Gardener any more.
+	if shoot.DeletionTimestamp != nil && !controllerutils.HasFinalizer(shoot, gardencorev1beta1.GardenerName) {
+		return controllerutils.PatchRemoveFinalizers(r.ctx, r.gardenClient, shoot, FinalizerName)
+	}
+
+	// Add finalizer to user managed and referenced secrets.
```

Comment:
```suggestion
	// Add finalizer to referenced secrets that are not managed by Gardener.
```

To make it more clear that finalizers are added to referenced secrets if they are not managed by Gardener. 
First, I thought that the finalizers are added to two different set of secrets (1) user managed 2) referenced secrets).